### ISSUE TITLE
Is there a way to use images with no bounding boxes to train a FasterRCNN?

### ISSUE THREAD

**Issue by joaqo** [pytorch/vision#4124]
Hi, I am currently using version 1.7.1 of pytorch with the corresponding version of torchvision. Everything is going perfect while training a FasterRCNN FPN 50 using your reference training project. That is until I added images with no bounding boxes to the training data, then the whole thing  broke down.

I need background images with no bounding boxes because the model needs to learn that certain types of images will not have objects in them, and if I only feed it images with objects in them, it will never see these types of images and predict nonsense when it gets them during inference.

Is this feature supported? The error I currently get when trying to train is:

```python
/home/joaqo/.local/lib/python3.6/site-packages/torchvision/models/detection/generalized_rcnn.py 69 forward
boxes.shape))

ValueError:
Expected target boxes to be a tensorof shape [N, 4], got torch.Size([0]).
```

Which makes perfect sense if you want to force your model to only train on images with targets, regrettably thats not my use case. If this is not currently supported? Would anyone provide some pointers for me to add support for this?

Thanks.

**Comment by oke-aditya** [comment]
Hi @joaqo !

Yes it is possible to use images with no bounding boxes in training data with torchvision.
This feature was implemented from release 0.6 
[Here](https://github.com/pytorch/vision/releases/tag/v0.6.0) is a note on how to use it.

If you have background image with No bounding boxes at all. The boxes are assumed to be a [0, 0, 0, 0], hence the area is 0. Also the label is assumed as background label which is 0 (predefined by torchvision models).

For e.g if such image has `image_id` =  3. So for such images you need to pass.

{
"image_id" : 3,
"labels": torch.zeros(0, dtype=torch.int64),
"boxes": torch.zeros((0, 4), dtype=torch.float32,
"area": torch.zeros(0, dtype=torch.float32),
}

I think the reference script will work out of box, if you make changes to your dataset.

Feel free to post problems you face in this issue.

**Comment by joaqo** [comment]
Hi @oke-aditya thanks for your prompt reply!

I've followed your instructions, and they work great when training on a simple gpu, but fail when using `torch.distributed.launch` for multi gpu training. I get the following error message:

```python
simple_detector/train.py 273 <module>
main(args)

simple_detector/train.py 141 main
train_one_epoch(model, optimizer, data_loader, device, epoch, args.print_freq, experiment_logger)

/home/joaqo/pano/simple_detector/engine.py 34 train_one_epoch
loss_dict = model(images, targets)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/modules/module.py 727 _call_impl
result = self.forward(*input, **kwargs)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/distributed.py 617 forward
inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/distributed.py 643 scatter
return scatter_kwargs(inputs, kwargs, device_ids, dim=self.dim)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 36 scatter_kwargs
inputs = scatter(inputs, target_gpus, dim) if inputs else []

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 28 scatter
res = scatter_map(inputs)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 15 scatter_map
return list(zip(*map(scatter_map, obj)))

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 17 scatter_map
return list(map(list, zip(*map(scatter_map, obj))))

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 19 scatter_map
return list(map(type(obj), zip(*map(scatter_map, obj.items()))))

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 15 scatter_map
return list(zip(*map(scatter_map, obj)))

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/scatter_gather.py 13 scatter_map
return Scatter.apply(target_gpus, None, dim, obj)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/_functions.py 92 forward
outputs = comm.scatter(input, target_gpus, chunk_sizes, ctx.dim, streams)

/home/joaqo/.local/lib/python3.6/site-packages/torch/nn/parallel/comm.py 186 scatter
return tuple(torch._C._scatter(tensor, devices, chunk_sizes, dim, streams))

RuntimeError:
chunk expects at least a 1-dimensional tensor
```

Not sure what could be the issue.

**Comment by oke-aditya** [comment]
Well I don't have any experience in multi-gpu training environments. Also, I don't have resources to reproduce this.

Here are few links which might help you.
I searched for error `chunk expects at least a 1-dimensional tensor`

https://discuss.pytorch.org/t/dataparallel-multi-gpu-runtimeerror-chunk-expects-at-least-a-1-dimensional-tensor/45340
https://stackoverflow.com/questions/56177305/dataparallel-multi-gpu-runtimeerror-chunk-expects-at-least-a-1-dimensional-tens

**Comment by joaqo** [comment]
I seem  to have fixed by inserting all the zeros in the example you linked to into arrays.

So from this:

```python
target = {"boxes": torch.zeros((0, 4), dtype=torch.float32),
          "labels": torch.zeros(0, dtype=torch.int64),
          "image_id": 4,
          "area": torch.zeros(0, dtype=torch.float32),
          "iscrowd": torch.zeros((0,), dtype=torch.int64)}
```

to this:

```python
target = {"boxes": torch.zeros((0, 4), dtype=torch.float32),
          "labels": torch.zeros([0], dtype=torch.int64),
          "image_id": torch.tensor([4]),  # Yeah, I even had to do it to image_id!
          "area": torch.zeros([0], dtype=torch.float32),
          "iscrowd": torch.zeros([0], dtype=torch.int64)}
```

Thanks!

**Comment by NicolasHug** [comment]
@joaqo could you provide a minimal reproducing example that works on GPU but fails on distributed? This could be a bug

**Comment by joaqo** [comment]
Hi @NicolasHug, yeah I'm pretty sure its a bug on Pytorch's side. The code I'm using is the [reference detection project](https://github.com/pytorch/vision/tree/master/references/detection) (Using pytorch 1.7.1) adapted to a different dataset. I think you should be able to reproduce it by setting the dataset to COCO, which is the default, and making sure the zeros are not inside an array, as shown in my previous comment.

It could also just be that the reference documentation is incorrect and in fact the model expects the `torch.zeros` array to be initialized with an array instead of just a plan `0`.

**Comment by sahilg06** [comment]
Hi @oke-aditya, does the use of images with no bounding boxes in training help? As we are only passing empty tensors in the target!

**Comment by oke-aditya** [comment]
Hi @sahilg06 .

Yes passing images with empty tensors do help in detection!! And this is a really good trick to improve detection. 
I will answer this on both ways, high level intuition so to know why and where it works. Also a precise answer to how it works 
:smile: Note that this method is general to object detection task, and should work out of box for many models.

Here is how! 

High Level intuition : -

Say you are detecting fruits (E.g. classes are apple, mango, orange)
And you have an image which contains, red tomatoes (I'm not considering tomato as fruit here).
If you have trained your model only on fruits images ignoring tomatoes then it is highly likely that a detection model will get confused with apples and tomatoes (since both are red, have similar shape,). The model never knew about tomato ! So it will make a biased guess that it is an apple!

Now, consider the case where you explicitly labelled these data as 0 tensors. The tomatoes now become a background object!
While training, the model now knows that the tomatoes are not apples! These are to be ignored. So it will learn better and hence differentiate tomatoes and apples much better. This method of marking the object as background class would help the model in reducing false positives.

Technical Answer: -

A detection problems involves 2 stages, classification and regression. It usually has a loss function which consists of 2 parts, classification loss and regression loss. By passing an object as background class, it will help the classification loss to differentiate target from background. When we explicitly pass background an object as background, these samples too help in determining classification loss.

Here is to the regression part!. By passing a tensor [0,0,0,0] for bounding box, this marks the object as a point object (basically a dot) which implies this is not a detectable object. So hence this is also ignored.

And when we do backward on these losses! This information gets propagated and hence the detector recognizes such explicit background objects.

(My PyTorch experience is far less, and hence I cannot provide a benchmark / any comparative analysis.)

**Comment by sahilg06** [comment]
Thanks, @oke-aditya  

But when we train for an image that contains bounding boxes (say image contains only 1 bbox), we pass the label as something like this [1], now model can learn that it is of 1st class. 

But when we pass the label for an image that doesn't contain any bounding box, we pass something like this [] (basically an empty tensor so even if pass `torch.ones(0, dtype=torch.int64)` instead of `torch.zeros(0, dtype=torch.int64)` it doesn't matter). So How model learns from this that it is of 0 or background class?

**Comment by joaqo** [comment]
Internally FasterRCNN (and many other object detection models) separate the image into areas called anchors, for each of these anchors they predict if it is one of the foreground classes or the background class. When training the model grabs a bunch of these anchors in different strategic ways, tries to balance a decent propotion between foreground and background classes (look [here](https://github.com/pytorch/vision/blob/master/torchvision/models/detection/faster_rcnn.py#L96) and [here](https://github.com/pytorch/vision/blob/master/torchvision/models/detection/faster_rcnn.py#L77) for the proportions in FasterRCNN), and calculates the loss function using them.

So it's trying to learn how well it did on the anchors it predicted as foreground but also on the ones it predicted as background. So you'll see that training on an image with no annotation is not really a problem, its as if you were just setting the proportion of foreground/background to just be background during that one sample. You cannot do it for all samples, but for a subset of them its fine.

**Comment by sahilg06** [comment]
Thanks, @joaqo, @oke-aditya. My exact doubt got cleared from this section of the code.https://github.com/pytorch/vision/blob/993325dd82567f5d4f28ccb321e3a9a16984d2d8/torchvision/models/detection/roi_heads.py#L569-L577

**Comment by fmassa** [comment]
Closing as seems that this has been answered / resolved already.
